### PR TITLE
add inference

### PR DESCRIPTION
**Pull request by caoxixiya**
running inference episodes in main function seems to be redundancy. I think it can be annotated.

### CODE REVIEW

File: examples/DDPG/train.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
 import gym
 import argparse
 import numpy as np
+import paddle
```
Comment (rical730) [inline examples/DDPG/train.py]: please remove `import paddle`

File: examples/DQN/train.py [hunk 2]
```diff
@@ -17,7 +17,8 @@
 import numpy as np
 import parl
 from parl.utils import logger, ReplayMemory
-
+import paddle
```
Comment (rical730) [inline examples/DQN/train.py]: please remove `import paddle`

File: examples/DDPG/train.py [hunk 3]
```diff
@@ -81,6 +83,23 @@ def run_evaluate_episodes(agent, env, eval_episodes):
     return avg_reward
 
 
+# Runs policy for 5 episodes by default and returns average reward
+# A fixed seed is used for the infer environment
+def run_inference_episodes(path, env, inference_episodes):
```
Comment (rical730) [inline examples/DDPG/train.py]: please use `run_evaluate_episodes()` to replace this function as we discuss

File: examples/DQN/train.py [hunk 4]
```diff
@@ -74,6 +75,27 @@ def run_evaluate_episodes(agent, env, eval_episodes=5, render=False):
     return np.mean(eval_reward)
 
 
+# infer 5 episodes
+def run_inference_episodes(path, env, inference_episodes=5, render=False):
```
Comment (rical730) [inline examples/DQN/train.py]: please use `run_evaluate_episodes()` to replace this function as we discuss

File: examples/DQN/train.py [hunk 5]
```diff
@@ -93,7 +115,7 @@ def main():
     while len(rpm) < MEMORY_WARMUP_SIZE:
         run_train_episode(agent, env, rpm)
 
-    max_episode = 800
+    max_episode = 80
```
Comment (rical730) [inline examples/DQN/train.py]: Please don't modify this value

File: parl/core/paddle/agent.py [hunk 6]
```diff
@@ -105,6 +102,29 @@ def save(self, save_path, model=None):
             model = self.alg.model
         paddle.save(model.state_dict(), save_path)
 
+    def save_inference_model(self, save_path, input_spec, model=None):
```
Comment (rical730) [inline parl/core/paddle/agent.py]: please add unittest for this API
refer1：https://github.com/PaddlePaddle/PARL/blob/develop/parl/core/paddle/tests/agent_base_test_paddle.py#L89 
refer2：https://github.com/PaddlePaddle/PARL/blob/develop/parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py#L113

File: examples/DDPG/mujoco_agent.py [hunk 7]
```diff
@@ -51,3 +51,7 @@ def learn(self, obs, action, reward, next_obs, terminal):
         critic_loss, actor_loss = self.alg.learn(obs, action, reward, next_obs,
                                                  terminal)
         return critic_loss, actor_loss
+
+    @staticmethod
+    def load_inference(path):
+        return paddle.jit.load(path)
```
Comment (rical730) [inline examples/DDPG/mujoco_agent.py]: please remove load_inference()

File: examples/DDPG/train.py [hunk 8]
```diff
@@ -122,6 +122,13 @@ def main():
             logger.info('Evaluation over: {} episodes, Reward: {}'.format(
                 EVAL_EPISODES, avg_reward))
 
+    # save the model and parameters of policy network for inference
+    save_inference_path = './inference_model'
+    input_shape = [[None, env.observation_space.shape[0]]]
+    input_type = ['float32']
+    agent.save_inference_model(save_inference_path, input_shape, input_type,
+                               model.actor_model)
```
Comment (rical730) [inline examples/DDPG/train.py:130]: input_shape -> input_shapes
input_type -> input_dtypes

File: examples/DQN/train.py [hunk 9]
```diff
@@ -112,6 +111,13 @@ def main():
     save_path = './model.ckpt'
     agent.save(save_path)
 
+    # save the model and parameters of policy network for inference
+    save_inference_path = './inference_model'
+    input_shape = [[None, env.observation_space.shape[0]]]
+    input_type = ['float32']
+    agent.save_inference_model(save_inference_path, input_shape, input_type,
+                               model)
```
Comment (rical730) [inline examples/DQN/train.py]: input_shape -> input_shapes
input_type -> input_dtypes
agent.save_inference_model(save_inference_path, input_shapes, input_dtypes)

File: parl/core/paddle/agent.py [hunk 10]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
             model = self.alg.model
         paddle.save(model.state_dict(), save_path)
 
+    def save_inference_model(self,
+                             save_path,
+                             input_shape_list,
+                             input_type_list,
```
Comment (rical730) [inline parl/core/paddle/agent.py]: input_type_list -> input_dtype_list

File: parl/core/paddle/agent.py [hunk 11]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+                             model=None):
+        """
+        Saves input Layer or function as ``paddle.jit.TranslatedLayer`` format model, which can be used for inference.
+
+        Args:
+            save_path(str): where to save the parameters.
+            model(parl.Model): model that describes the policy network structure. If None, will use self.alg.model.
+            input_shape_list(list): shape of all inputs of the saved model's forward method.
+            input_shape_type(list): dtype of all inputs of the saved model's forward method.
```
Comment (rical730) [inline parl/core/paddle/agent.py]: input_shape_type -> input_dtype_list

File: parl/core/paddle/agent.py [hunk 12]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+    def save_inference_model(self,
+                             save_path,
+                             input_shape_list,
+                             input_type_list,
+                             model=None):
+        """
+        Saves input Layer or function as ``paddle.jit.TranslatedLayer`` format model, which can be used for inference.
+
+        Args:
+            save_path(str): where to save the parameters.
```
Comment (rical730) [inline parl/core/paddle/agent.py]: where to save the inference_model.

File: parl/core/paddle/agent.py [hunk 13]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+            model(parl.Model): model that describes the policy network structure. If None, will use self.alg.model.
+            input_shape_list(list): shape of all inputs of the saved model's forward method.
+            input_shape_type(list): dtype of all inputs of the saved model's forward method.
+
+        Example:
+
+        .. code-block:: python
+
+            agent = AtariAgent()
+            agent.save_inference_model('./model_dir', [[None, 128]], ['float32'], model)
```
Comment (rical730) [inline parl/core/paddle/agent.py]: agent.save_inference_model('./model_dir', [[None, 128]], ['float32'])

File: parl/core/paddle/agent.py [hunk 14]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+
+
+        Example with actor-critic:
+
+        .. code-block:: python
+
+            agent = AtariAgent()
+            agent.save_inference_model('./model_dir', [[None, 128]], ['float32'], model.actor_model)
```
Comment (rical730) [inline parl/core/paddle/agent.py]: agent.save_inference_model('./model_dir', [[None, 128]], ['float32'], agent.alg.model.actor_model)

File: parl/core/paddle/agent.py [hunk 15]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+
+        """
+        if model is None:
+            model = self.alg.model
+        assert hasattr(model, 'forward'), "model must have forward method"
+        assert isinstance(input_shape_list, list), "input_shape must be list"
```
Comment (rical730) [inline parl/core/paddle/agent.py]: 'Type of `input_shape_list` in save_inference_model() should be list, but received {}'.format(type(input_shape_list)).

File: parl/core/paddle/agent.py [hunk 16]
```diff
@@ -105,6 +103,47 @@ def save(self, save_path, model=None):
...
+        assert isinstance(input_type_list, list), "input_type must be list"
```
Comment (rical730) [inline parl/core/paddle/agent.py]: please update error message refer to `input_shape_list`

File: parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py [hunk 17]
```diff
@@ -119,6 +148,35 @@ def test_save(self):
         self.assertTrue(os.path.exists(save_path1))
         self.assertTrue(os.path.exists(save_path2))
 
+    def test_inference_model(self):
+        agent = TestAgent(self.alg)
+        save_path1 = 'my_acmodel'
+        save_path2 = os.path.join('my_ac_model', 'model-2')
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py]: save_path1 = 'my_inference_model'
save_path2 = os.path.join('my_infer_models', 'model-2')

File: parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py [hunk 18]
```diff
@@ -119,6 +148,35 @@ def test_save(self):
...
+        input_spec = [
+            paddle.static.InputSpec(shape=[None, 4], dtype='float32')
+        ]
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py]: please remove `input_spec`

File: parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py [hunk 19]
```diff
@@ -119,6 +148,35 @@ def test_save(self):
         self.assertTrue(os.path.exists(save_path1))
         self.assertTrue(os.path.exists(save_path2))
 
+    def test_inference_model(self):
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py]: test_inference_model ->  test_save_inference_model

File: parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py [hunk 20]
```diff
@@ -119,6 +148,35 @@ def test_save(self):
...
+        input_shape = [[None, 4]]
+        input_type = ['float32']
+        agent.save_inference_model(save_path1, input_shape, input_type,
+                                   self.model.actor)
+        agent.save_inference_model(save_path2, input_shape, input_type,
+                                   self.model.actor)
+        self.assertTrue(os.path.exists(save_path1 + '.pdmodel'))
+        self.assertTrue(os.path.exists(save_path2 + '.pdmodel'))
+
+    def test_double_inference_model(self):
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py]: test_double_inference_model -> test_save_inference_model_with_multi_inputs

File: parl/core/paddle/tests/agent_base_test_paddle.py [hunk 21]
```diff
@@ -95,6 +117,32 @@ def test_save(self):
         self.assertTrue(os.path.exists(save_path1))
         self.assertTrue(os.path.exists(save_path2))
 
+    def test_inference_model(self):
+        agent = TestAgent(self.alg)
+        save_path1 = 'my_model'
+        save_path2 = os.path.join('my_model', 'model-2')
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_test_paddle.py]: save_path1 = 'my_inference_model'
save_path2 = os.path.join('my_infer_models', 'model-2')

File: parl/core/paddle/tests/agent_base_test_paddle.py [hunk 22]
```diff
@@ -95,6 +117,32 @@ def test_save(self):
         self.assertTrue(os.path.exists(save_path1))
         self.assertTrue(os.path.exists(save_path2))
 
+    def test_inference_model(self):
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_test_paddle.py]: test_inference_model -> test_save_inference_model

File: parl/core/paddle/tests/agent_base_test_paddle.py [hunk 23]
```diff
@@ -95,6 +117,32 @@ def test_save(self):
...
+        input_shape = [[None, 4]]
+        input_type = ['float32']
+        agent.save_inference_model(save_path1, input_shape, input_type,
+                                   self.model)
+        agent.save_inference_model(save_path2, input_shape, input_type,
+                                   self.model)
+        self.assertTrue(os.path.exists(save_path1 + '.pdmodel'))
+        self.assertTrue(os.path.exists(save_path2 + '.pdmodel'))
+
+    def test_double_inference_model(self):
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_test_paddle.py]: test_double_inference_model -> test_save_inference_model_with_multi_inputs

File: parl/core/paddle/tests/agent_base_test_paddle.py (hunk 22 above)
Comment (rical730) [reply parl/core/paddle/tests/agent_base_test_paddle.py]: please test AssertError situation refer to https://github.com/PaddlePaddle/PARL/blob/develop/parl/core/paddle/tests/agent_base_actor_critic_test_paddle.py#L247

File: parl/core/paddle/tests/agent_base_test_paddle.py [hunk 24]
```diff
@@ -27,19 +27,47 @@ def __init__(self):
...
+        out = self.fc3(out)
+        return out
+
+
+class TestModelWithoutForward(parl.Model):
+    def __init__(self):
+        super(TestModelWithoutForward, self).__init__()
+        self.fc1 = nn.Linear(4, 256)
+        self.fc2 = nn.Linear(256, 128)
+        self.fc3 = nn.Linear(128, 1)
```
Comment (rical730) [inline parl/core/paddle/tests/agent_base_test_paddle.py:42]: please add `self.forward = None` for testing line 138 of parl/core/paddle/agent.py

**Review by TomorrowIsAnOtherDay — DISMISSED** [review]
LGTM.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]

**Pull request merged**